From 2f4fb3842ebf171b5124d9c23e1b319f79dc6e73 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sun, 23 Mar 2008 19:28:59 +0000
Subject: [PATCH] CONFIGURATION-316: The text of the root element of an
 XMLConfiguration is now explicitly set when the configuration is saved.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@640240 13f79535-47bb-0310-9956-ffa450edef68
---
 .../configuration/XMLConfiguration.java       | 31 ++++++++++++++++++-
 .../configuration/TestXMLConfiguration.java   | 30 ++++++++++++++++++
 xdocs/changes.xml                             |  4 +++
 3 files changed, 64 insertions(+), 1 deletion(-)

diff --git a/src/java/org/apache/commons/configuration/XMLConfiguration.java b/src/java/org/apache/commons/configuration/XMLConfiguration.java
index 041fb2f7e9..bfa6b702f7 100644
--- a/src/java/org/apache/commons/configuration/XMLConfiguration.java
+++ b/src/java/org/apache/commons/configuration/XMLConfiguration.java
@@ -620,8 +620,9 @@ protected Document createDocument() throws ConfigurationException
             XMLBuilderVisitor builder = new XMLBuilderVisitor(document,
                     isDelimiterParsingDisabled() ? (char) 0 : getListDelimiter());
             builder.processDocument(getRoot());
+            initRootElementText(document, getRootNode().getValue());
             return document;
-        } /* try */
+        }
         catch (DOMException domEx)
         {
             throw new ConfigurationException(domEx);
@@ -632,6 +633,34 @@ protected Document createDocument() throws ConfigurationException
         }
     }
 
+    /**
+     * Sets the text of the root element of a newly created XML Document.
+     *
+     * @param doc the document
+     * @param value the new text to be set
+     */
+    private void initRootElementText(Document doc, Object value)
+    {
+        Element elem = doc.getDocumentElement();
+        NodeList children = elem.getChildNodes();
+
+        // Remove all existing text nodes
+        for (int i = 0; i < children.getLength(); i++)
+        {
+            org.w3c.dom.Node nd = children.item(i);
+            if (nd.getNodeType() == org.w3c.dom.Node.TEXT_NODE)
+            {
+                elem.removeChild(nd);
+            }
+        }
+
+        if (value != null)
+        {
+            // Add a new text node
+            elem.appendChild(doc.createTextNode(String.valueOf(value)));
+        }
+    }
+
     /**
      * Creates a new node object. This implementation returns an instance of the
      * <code>XMLNode</code> class.
diff --git a/src/test/org/apache/commons/configuration/TestXMLConfiguration.java b/src/test/org/apache/commons/configuration/TestXMLConfiguration.java
index ac379d31c7..1e1688a14e 100644
--- a/src/test/org/apache/commons/configuration/TestXMLConfiguration.java
+++ b/src/test/org/apache/commons/configuration/TestXMLConfiguration.java
@@ -997,6 +997,36 @@ public void testInitCopy() throws ConfigurationException
         checkSavedConfig(copy);
     }
 
+    /**
+     * Tests setting text of the root element.
+     */
+    public void testSetTextRootElement() throws ConfigurationException
+    {
+        conf.setProperty("", "Root text");
+        conf.save(testSaveConf);
+        XMLConfiguration copy = new XMLConfiguration();
+        copy.setFile(testSaveConf);
+        checkSavedConfig(copy);
+    }
+
+    /**
+     * Tests removing the text of the root element.
+     */
+    public void testClearTextRootElement() throws ConfigurationException
+    {
+        final String xml = "<e a=\"v\">text</e>";
+        conf.clear();
+        StringReader in = new StringReader(xml);
+        conf.load(in);
+        assertEquals("Wrong text of root", "text", conf.getString(""));
+
+        conf.clearProperty("");
+        conf.save(testSaveConf);
+        XMLConfiguration copy = new XMLConfiguration();
+        copy.setFile(testSaveConf);
+        checkSavedConfig(copy);
+    }
+
     /**
      * Tests list nodes with multiple values and attributes.
      */
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index 555899ecb2..19d2211f43 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -23,6 +23,10 @@
 
   <body>
     <release version="1.6" date="in SVN" description="">
+      <action dev="oheger" type="fix" issue="CONFIGURATION-316">
+        Changing the text of the root element of an XMLConfiguration had no
+        effect when the configuration was saved. This has been fixed.
+      </action>
       <action dev="oheger" type="fix" issue="CONFIGURATION-315">
         CombinedConfiguration used to send two EVENT_COMBINED_INVALIDATE events
         for each modified child configuration. Now this event is sent only
